File: benchmarks/erlang/binary_trees.erl
-module(binary_trees).
-export([main/0]).

% A tree is represented as {item, Left, Right} or nil

% Create a new tree with the given item value at the root
new_tree(Item, 0) ->
    {Item, nil, nil};
new_tree(Item, Depth) ->
    {Item, new_tree(2*Item-1, Depth-1), new_tree(2*Item, Depth-1)}.

% Check the tree and return a checksum
check_tree(nil) ->
    0;
check_tree({Item, nil, nil}) ->
    Item;
check_tree({Item, Left, Right}) ->
    Item + check_tree(Left) - check_tree(Right).

% Process multiple trees
process_trees(_, _, 0, Result) ->
    Result;
process_trees(Depth, I, Remaining, Result) ->
    A = new_tree(I, Depth),
    B = new_tree(-I, Depth),
    NewResult = Result + check_tree(A) + check_tree(B),
    process_trees(Depth, I+1, Remaining-1, NewResult).

main() ->
    MinDepth = 4,
    MaxDepth = 12,
    StretchDepth = MaxDepth + 1,
    StartTime = erlang:system_time(millisecond),
    
    % Allocate and check a big tree
    BigTree = new_tree(0, StretchDepth),
    io:format("stretch tree of depth ~p check: ~p~n", [StretchDepth, check_tree(BigTree)]),
    
    % Allocate a long-lived binary tree
    LongLivedTree = new_tree(0, MaxDepth),
    
    % Check trees of increasing depth
    depths_loop(MinDepth, MaxDepth, MinDepth),
    
    % Check the long-lived tree last
    io:format("long lived tree of depth ~p check: ~p~n", [MaxDepth, check_tree(LongLivedTree)]),
    
    ElapsedTime = erlang:system_time(millisecond) - StartTime,
    io:format("Time taken: ~p ms~n", [ElapsedTime]),
    
    {ok, [TotalWords, _]} = process_info(self(), memory),
    io:format("Memory used: ~p KB~n", [TotalWords div 1024]),
    ok.

depths_loop(Depth, MaxDepth, _) when Depth > MaxDepth ->
    ok;
depths_loop(Depth, MaxDepth, MinDepth) ->
    Iterations = 1 bsl (MaxDepth - Depth + MinDepth),
    Result = process_trees(Depth, 0, Iterations, 0),
    io:format("~p trees of depth ~p check: ~p~n", [Iterations*2, Depth, Result]),
    depths_loop(Depth+2, MaxDepth, MinDepth).

File: benchmarks/ocaml/binary_trees.ml
(* Binary trees benchmark adapted from The Computer Language Benchmarks Game *)

(* A tree type *)
type tree = 
  | Nil
  | Node of int * tree * tree

(* Create a new tree with the given item value at the root *)
let rec new_tree item depth =
  if depth <= 0 then
    Node(item, Nil, Nil)
  else
    Node(item, 
         new_tree (2 * item - 1) (depth - 1),
         new_tree (2 * item) (depth - 1))

(* Check the tree and return a checksum *)
let rec check_tree = function
  | Nil -> 0
  | Node(item, Nil, Nil) -> item
  | Node(item, left, right) -> item + check_tree left - check_tree right

(* Process multiple trees *)
let rec process_trees depth i remaining result =
  if remaining = 0 then
    result
  else
    let a = new_tree i depth in
    let b = new_tree (-i) depth in
    let new_result = result + check_tree a + check_tree b in
    process_trees depth (i + 1) (remaining - 1) new_result

let () =
  let min_depth = 4 in
  let max_depth = 12 in
  let stretch_depth = max_depth + 1 in
  let start_time = Unix.gettimeofday () in
  
  (* Allocate and check a big tree *)
  let stretch_tree = new_tree 0 stretch_depth in
  Printf.printf "stretch tree of depth %d check: %d\n" stretch_depth (check_tree stretch_tree);
  
  (* Allocate a long-lived binary tree *)
  let long_lived_tree = new_tree 0 max_depth in
  
  (* Check trees of increasing depth *)
  let rec depth_loop depth =
    if depth <= max_depth then begin
      let iterations = 1 lsl (max_depth - depth + min_depth) in
      let result = process_trees depth 0 iterations 0 in
      Printf.printf "%d trees of depth %d check: %d\n" (iterations * 2) depth result;
      depth_loop (depth + 2)
    end
  in
  depth_loop min_depth;
  
  (* Check the long-lived tree last *)
  Printf.printf "long lived tree of depth %d check: %d\n" max_depth (check_tree long_lived_tree);
  
  let elapsed = int_of_float ((Unix.gettimeofday () -. start_time) *. 1000.0) in
  Printf.printf "Time taken: %d ms\n" elapsed;
  
  (* Approximate memory usage *)
  let memory_usage = Gc.stat () in
  Printf.printf "Memory used: %d KB\n" (memory_usage.minor_words |> int_of_float |> (fun x -> x / 1024))